### PR TITLE
Implementations of atomic and merged algorithms for weighted average wirelength computations on CPU

### PR DESCRIPTION
**Pull request by ZixuanJiang**
We follow the GPU implementations and complete the CPU implementations of atomic (pin by pin) and merged weighted average wirelength.

Since atomic min and max operations are not supported by OpenMP, we use `computeMaxMinNetByNet` instead of `PinByPin` in the atomic algorithm on CPU. That is the only difference between CPU and GPU implementations.

Authors: Zixuan Jiang and Jiaqi Gu.

### CODE REVIEW

**Comment by ZixuanJiang** [comment]
Do we decide to support torch 1.3.1?

**Comment by limbo018** [comment]
> Do we decide to support torch 1.3.1?

I saw you add something like ```padding_mask.to(at::ScalarType::Bool)``` for torch 1.3.1, but I got errors when compiling under 1.0.0. Do you have better way to resolve this issue for ```masked_fill_``` API? Probably change the tensor type in Python is better than in C++.

**Comment by limbo018** [comment]
> Do we decide to support torch 1.3.1?

I have merged your code to routability branch and will then merge to develop and master after more test.

**Comment by ZixuanJiang** [comment]
There is only one commit that is relevant to the torch 1.3.1.
https://github.com/limbo018/DREAMPlace/pull/13/commits/42dfdbc6def43125fc546c8edf9036283924d2fc
Revert this commit if we plan to stay in torch 1.0.

File: dreamplace/ops/density_potential/src/density_potential_cuda.cpp [hunk 1]
```diff
@@ -172,7 +172,7 @@ std::vector<at::Tensor> density_potential_forward(
     // set padding density 
     if (padding > 0)
     {
-        density_map.masked_fill_(padding_mask, at::Scalar(target_area));
+        density_map.masked_fill_(padding_mask.to(at::ScalarType::Bool), at::Scalar(target_area));
```
Comment (limbo018) [inline dreamplace/ops/density_potential/src/density_potential_cuda.cpp]: Can you find equivalent python API for ```padding_mask.to(at::ScalarType::Bool)```? I am more comfortable to do it in python.

File: dreamplace/ops/electric_potential/src/electric_density_map.cpp [hunk 2]
```diff
@@ -167,7 +167,7 @@ at::Tensor density_map(
     // set padding density
     if (padding > 0)
     {
-        density_map.masked_fill_(padding_mask, at::Scalar(target_density*bin_size_x*bin_size_y));
+        density_map.masked_fill_(padding_mask.to(at::ScalarType::Bool), at::Scalar(target_density*bin_size_x*bin_size_y));
```
Comment (limbo018) [inline dreamplace/ops/electric_potential/src/electric_density_map.cpp]: replace it with a python API.

File: dreamplace/ops/electric_potential/src/electric_density_map_cuda.cpp [hunk 3]
```diff
@@ -169,7 +169,7 @@ at::Tensor density_map(
     // set padding density 
     if (padding > 0)
     {
-        density_map.masked_fill_(padding_mask, at::Scalar(target_density*bin_size_x*bin_size_y));
+        density_map.masked_fill_(padding_mask.to(at::ScalarType::Bool), at::Scalar(target_density*bin_size_x*bin_size_y));
```
Comment (limbo018) [inline dreamplace/ops/electric_potential/src/electric_density_map_cuda.cpp]: replace it with a python API.

File: dreamplace/ops/hpwl/src/hpwl_atomic.cpp [hunk 4]
```diff
@@ -52,10 +52,10 @@ at::Tensor hpwl_atomic_forward(
     at::Tensor partial_hpwl_min = at::zeros({2, num_nets}, pos.type()); 
 
     DREAMPLACE_DISPATCH_FLOATING_TYPES(pos.type(), "computeHPWLAtomicLauncher", [&] {
-            partial_hpwl_max[0].masked_fill_(net_mask, std::numeric_limits<scalar_t>::min());
-            partial_hpwl_max[1].masked_fill_(net_mask, std::numeric_limits<scalar_t>::min());
-            partial_hpwl_min[0].masked_fill_(net_mask, std::numeric_limits<scalar_t>::max());
-            partial_hpwl_min[1].masked_fill_(net_mask, std::numeric_limits<scalar_t>::max());
+            partial_hpwl_max[0].masked_fill_(net_mask.to(at::ScalarType::Bool), std::numeric_limits<scalar_t>::min());
```
Comment (limbo018) [inline dreamplace/ops/hpwl/src/hpwl_atomic.cpp]: Replace these ```net_mask.to``` with a python API.

File: dreamplace/ops/hpwl/src/hpwl_atomic.cpp [hunk 5]
```diff
@@ -52,10 +52,10 @@ at::Tensor hpwl_atomic_forward(
...
+            partial_hpwl_max[1].masked_fill_(net_mask.to(at::ScalarType::Bool), std::numeric_limits<scalar_t>::min());
```
Comment (limbo018) [inline dreamplace/ops/hpwl/src/hpwl_atomic.cpp]: Replace these ```net_mask.to``` with a python API.

File: dreamplace/ops/hpwl/src/hpwl_cuda_atomic.cpp [hunk 6]
```diff
@@ -53,10 +53,10 @@ at::Tensor hpwl_atomic_forward(
     at::Tensor scaled_pos = at::_cast_Int(pos.mul(1000), false);
     at::Tensor partial_hpwl_max = at::zeros({2, num_nets}, scaled_pos.type()); 
     at::Tensor partial_hpwl_min = at::zeros({2, num_nets}, scaled_pos.type()); 
-    partial_hpwl_max[0].masked_fill_(net_mask, std::numeric_limits<T>::min());
-    partial_hpwl_max[1].masked_fill_(net_mask, std::numeric_limits<T>::min());
-    partial_hpwl_min[0].masked_fill_(net_mask, std::numeric_limits<T>::max());
-    partial_hpwl_min[1].masked_fill_(net_mask, std::numeric_limits<T>::max());
+    partial_hpwl_max[0].masked_fill_(net_mask.to(at::ScalarType::Bool), std::numeric_limits<T>::min());
```
Comment (limbo018) [inline dreamplace/ops/hpwl/src/hpwl_cuda_atomic.cpp]: Replace these ```net_mask.to``` with a python API.

File: dreamplace/ops/density_potential/src/density_potential_cuda.cpp (hunk 1 above)
Comment (limbo018) [reply dreamplace/ops/density_potential/src/density_potential_cuda.cpp]: I have moved it to python.

File: dreamplace/ops/electric_potential/src/electric_density_map.cpp (hunk 2 above)
Comment (limbo018) [reply dreamplace/ops/electric_potential/src/electric_density_map.cpp]: I have moved it to python.

File: dreamplace/ops/hpwl/src/hpwl_atomic.cpp (hunk 4 above)
Comment (limbo018) [reply dreamplace/ops/hpwl/src/hpwl_atomic.cpp]: I have removed masked_fill and handle net_mask in C++/CUDA.

File: dreamplace/ops/hpwl/src/hpwl_atomic.cpp (hunk 5 above)
Comment (limbo018) [reply dreamplace/ops/hpwl/src/hpwl_atomic.cpp]: I have removed masked_fill_ and handle it in C++/CUDA.

File: dreamplace/ops/hpwl/src/hpwl_cuda_atomic.cpp (hunk 6 above)
Comment (limbo018) [reply dreamplace/ops/hpwl/src/hpwl_cuda_atomic.cpp]: I have removed masked_fill_ and handle it in C++/CUDA.

File: dreamplace/ops/electric_potential/src/electric_density_map_cuda.cpp (hunk 3 above)
Comment (limbo018) [reply dreamplace/ops/electric_potential/src/electric_density_map_cuda.cpp]: I have moved it to python.

**Comment by ZixuanJiang** [comment]
This pull request is merged in the `develop` branch. Therefore, we close it.

**Pull request closed without merging**